cdp/smart_contract.py

```diff
@@ -0,0 +1,324 @@
+        """Get the options of the smart contract.
+
+        Returns:
+            Union[TokenOptions, NFTOptions, MultiTokenOptions]: The smart contract options.
+
+        """
+        if self.type == self.Type.ERC20:
+            return self.TokenOptions(
+                name=self._model.options.name,
+                symbol=self._model.options.symbol,
+                total_supply=self._model.options.total_supply,
+            )
+        elif self.type == self.Type.ERC721:
+            return self.NFTOptions(
+                name=self._model.options.name,
+                symbol=self._model.options.symbol,
+                base_uri=self._model.options.base_uri,
+            )
+        else:
+            return self.MultiTokenOptions(uri=self._model.options.uri)
+
+    @property
+    def abi(self) -> dict[str, Any]:
+        """Get the ABI of the smart contract.
+
+        Returns:
+            Dict[str, Any]: The ABI as a JSON object.
+
+        """
+        return json.loads(self._model.abi)
+
+    @property
+    def transaction(self) -> Transaction:
+        """Get the transaction associated with the smart contract deployment.
+
+        Returns:
+            Transaction: The transaction.
+
+        """
+        if self._transaction is None and self._model.transaction is not None:
+            self._update_transaction(self._model)
+        return self._transaction
+
+    def sign(self, key: LocalAccount) -> "SmartContract":
+        """Sign the smart contract deployment with the given key.
+
+        Args:
+            key (LocalAccount): The key to sign the smart contract deployment with.
+
+        Returns:
+            SmartContract: The signed SmartContract object.
+
+        Raises:
+            ValueError: If the key is not a LocalAccount.
+
+        """
+        if not isinstance(key, LocalAccount):
+            raise ValueError("key must be a LocalAccount")
+
+        self.transaction.sign(key)
+        return self
+
+    def broadcast(self) -> "SmartContract":
+        """Broadcast the smart contract deployment to the network.
+
+        Returns:
+            SmartContract: The broadcasted SmartContract object.
+
+        Raises:
+            ValueError: If the smart contract deployment is not signed.
+
+        """
+        if not self.transaction.signed:
+            raise ValueError("Cannot broadcast unsigned SmartContract deployment")
+
+        deploy_smart_contract_request = DeploySmartContractRequest(
+            signed_payload=self.transaction.signature
+        )
+
+        model = Cdp.api_clients.smart_contracts.deploy_smart_contract(
+            wallet_id=self.wallet_id,
+            deployer_address=self.deployer_address,
+            smart_contract_id=self.smart_contract_id,
+            deploy_smart_contract_request=deploy_smart_contract_request,
+        )
+        self._model = model
+        return self
+
+    def reload(self) -> "SmartContract":
```

Comment:
We have been returning `None` on reload's since they are modifying the underlying object instance

Reply:
contract invocation returns self though?

```

    def reload(self) -> "ContractInvocation":
        """Reload the Contract Invocation model with the latest version from the server.

        Returns:
            ContractInvocation: The updated ContractInvocation object.

        """
        model = Cdp.api_clients.contract_invocations.get_contract_invocation(
            wallet_id=self.wallet_id,
            address_id=self.address_id,
            contract_invocation_id=self.contract_invocation_id,
        )

        self._model = model
        self._update_transaction(model)

        return self
        ```

should i update that to return None as well?

Reply:
Let's leave it as is and we can reassess in a separate pass of all `reload` instance methods.